### PR TITLE
Make our pattern matching logic undestand `&[T]` constants

### PR DESCRIPTION
**Pull request by oli-obk**
We should generalize this to other types, too, so treat this as a proof of concept.

### CODE REVIEW

File: src/librustc_mir_build/hair/pattern/_match.rs [hunk 1]
```diff
@@ -291,30 +291,39 @@ impl<'tcx> LiteralExpander<'tcx> {
...
+                        assert!(layout.is_zst());
+                        // Deref of a reference to a ZST is a nop.
+                        ConstValue::Scalar(Scalar::zst())
                     }
                 }
             }
-            // unsize array to slice if pattern is array but match value or other patterns are slice
-            (ConstValue::Scalar(Scalar::Ptr(p)), ty::Array(t, n), ty::Slice(u)) => {
+            // Unsize array to slice if pattern is array
+            // but match value or other patterns are slice.
```
Comment (varkor) [inline src/librustc_mir_build/hair/pattern/_match.rs:301]:
```suggestion
			// Unsize the array to a slice if we're matching on a slice,
			// or other patterns are a slice (despite this pattern being
			// an array).
```
I found the current comment a little hard to parse.

File: src/librustc_mir_build/hair/pattern/_match.rs [hunk 2]
```diff
@@ -291,30 +291,39 @@ impl<'tcx> LiteralExpander<'tcx> {
                     }
                     Scalar::Raw { .. } => {
                         let layout = self.tcx.layout_of(self.param_env.and(rty)).unwrap();
-                        if layout.is_zst() {
-                            // Deref of a reference to a ZST is a nop.
-                            ConstValue::Scalar(Scalar::zst())
-                        } else {
-                            // FIXME(oli-obk): this is reachable for `const FOO: &&&u32 = &&&42;`
```
Comment (varkor) [inline src/librustc_mir_build/hair/pattern/_match.rs:298]: Was this comment incorrect before?

File: src/librustc_mir_build/hair/pattern/_match.rs [hunk 3]
```diff
@@ -291,30 +291,39 @@ impl<'tcx> LiteralExpander<'tcx> {
...
+                        // Deref of a reference to a ZST is a nop.
+                        ConstValue::Scalar(Scalar::zst())
                     }
                 }
             }
-            // unsize array to slice if pattern is array but match value or other patterns are slice
-            (ConstValue::Scalar(Scalar::Ptr(p)), ty::Array(t, n), ty::Slice(u)) => {
+            // Unsize array to slice if pattern is array
+            // but match value or other patterns are slice.
+            (ConstValue::Scalar(s), ty::Array(t, n), ty::Slice(u)) => {
```
Comment (Centril) [inline src/librustc_mir_build/hair/pattern/_match.rs:303]: Would be good to avoid these one-letter variable names.

File: src/librustc_mir_build/hair/pattern/_match.rs (hunk 2 above)
Comment (oli-obk) [reply src/librustc_mir_build/hair/pattern/_match.rs:298]: apparently, we have a test for this, but the ICE is a different one somewhere in the middle of pattern matching. I've so far been unable to figure it out.

**Comment by oli-obk** [comment]
I found a rabbit hole and wrote some text about it: https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/constants.20in.20patterns/near/192783049

I want to fix this once and for all, instead of the piecemeal thing that I've been doing the last years.

**Comment by oli-obk** [comment]
closing in favour of https://github.com/rust-lang/rust/pull/70743
